### DISCUSSION TITLE
Is it possible to use distroless image?

### DISCUSSION THREAD

**Discussion by gregjotau** [basecamp/kamal#661] [General]
When I use a distroless image (deploying java/kotlin apps) Kamal fails due to health check not working, probably because the containers do not have curl available. 

Would it be possible for curl locally on my machine to be used to check for availability instead or allow to ignore the health check (make it optional)?

**Comment by acidtib** [comment]
is installing curl not possible?

**Reply by gregjotau** [reply]
installing additional packages such as curl is not super easy. Distroless images do not have package managers and does not include a shell. Also defeats a bit the purpose with distroless I guess :)

**Comment by pschmitt-tc** [comment]
@gregjotau 

I read that you can try to run the deploy without health check:

```bash
kamal deploy --skip-hooks
```

*This command will deploy your application to the servers without running the healthcheck or any other custom scripts you have set up in the hooks.*

**Reply by gregjotau** [reply]
kamal deploy --skip-hooks
-> 
  ERROR (Kamal::Cli::Healthcheck::Poller::HealthcheckError): Exception while executing on host oslo2: container not ready (unhealthy)

  "Log": [
    {
      "Start": "2024-01-25T15:42:19.558546639Z",
      "End": "2024-01-25T15:42:19.623512865Z",
      "ExitCode": -1,
      "Output": "OCI runtime exec failed: exec failed: unable to start container process: exec: \"/bin/sh\": stat /bin/sh: no such file or directory: unknown"
    },

Does not work. I think the keyword here is "other **customer** scripts". The built in health-check is not possible to skip.

**Reply by pedromschmitt** [reply]
Ok... This not work.

You can also mock the healthcheck answer, so Kamal will see it as true.


This code I tested before:

```
 healthcheck:
   cmd: ls -l > /dev/null
```

**Reply by gregjotau** [reply]
Ensure app can pass healthcheck...
  INFO [0491581f] Running docker run --detach --name healthcheck-ecomtools-3f0223e609839c9f7a264e55fde0caaeaa4c0dfc_uncommitted_58eee533750cc1ef --publish 3999:8080 --label service=healthcheck-ecomtools -e KAMAL_CONTAINER_NAME="healthcheck-ecomtools" --env-file .kamal/env/roles/ecomtools-web.env --health-cmd "ls -l > /dev/null" --health-interval "1s" gregtau/ecomtools:3f0223e609839c9f7a264e55fde0caaeaa4c0dfc_uncommitted_58eee533750cc1ef on oslo2
  INFO [0491581f] Finished in 0.791 seconds with exit status 0 (successful).
  INFO [bb191bd2] Running docker container ls --all --filter name=^healthcheck-ecomtools-3f0223e609839c9f7a264e55fde0caaeaa4c0dfc_uncommitted_58eee533750cc1ef$ --quiet | xargs docker inspect --format '{{if .State.Health}}{{.State.Health.Status}}{{else}}{{.State.Status}}{{end}}' on oslo2
  INFO [bb191bd2] Finished in 0.127 seconds with exit status 0 (successful).
  INFO container not ready (starting), retrying in 1s (attempt 1/7)...
  INFO [1fda577c] Running docker container ls --all --filter name=^healthcheck-ecomtools-3f0223e609839c9f7a264e55fde0caaeaa4c0dfc_uncommitted_58eee533750cc1ef$ --quiet | xargs docker inspect --format '{{if .State.Health}}{{.State.Health.Status}}{{else}}{{.State.Status}}{{end}}' on oslo2
  INFO [1fda577c] Finished in 0.086 seconds with exit status 0 (successful).
  INFO container not ready (starting), retrying in 2s (attempt 2/7)...


  ERROR (Kamal::Cli::Healthcheck::Poller::HealthcheckError): Exception while executing on host oslo2: container not ready (unhealthy)

      "Output": "OCI runtime exec failed: exec failed: unable to start container process: exec: \"/bin/sh\": stat /bin/sh: no such file or directory: unknown"

Did you test this on a distroless image? Doesn't seem to work :(

**Reply by pschmitt-tc** [reply]
Not a distroless...
And this error makes sense - I just googled:

`"Distroless" images contain only your application and its runtime dependencies. They do not contain package managers, shells or any other programs`

Without a shell we cant send any command like this `ls` to there...

**Reply by acidtib** [reply]
--skip-hooks prevent hook scripts from running, it has nothing to do with healthcheck 

if you want to fake helathcheck success add a simple .sh file that returns true and call it using the healthcheck config

```
 healthcheck:
   cmd: ./bin/good_check.sh
```

**Reply by Sija** [reply]
Wouldn't simply using `cmd: true` be enough?

**Reply by acidtib** [reply]
you can try it, i dont think it will work, the value of `cmd:` is passed to docker run param `--health-cmd` you can see some examples here in the healcheck test

https://github.com/basecamp/kamal/blob/aea55480adcaf61e3eebfd49f2b8c039207ad0de/test/commands/healthcheck_test.rb#L37
